cpp/include/cudf/strings/regex/regex_program.hpp

```diff
@@ -0,0 +1,138 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/regex/flags.hpp>
+#include <cudf/types.hpp>
+
+#include <memory>
+#include <string>
+
+namespace cudf {
+namespace strings {
+
+/**
+ * @addtogroup strings_regex
+ * @{
+ */
+
+/**
+ * @brief Regex program class
+ *
+ * Create an instance from a regex pattern and use it to call the appropriate
+ * strings APIs. An instance can be reused.
+ *
+ * See the @ref md_regex "Regex Features" page for details on patterns and APIs the support regex.
+ */
+struct regex_program {
+  struct regex_program_impl;
+
+  /**
+   * @brief Create a program from a pattern
+   *
+   * @throw cudf::logic_error If pattern is found to be invalid or contain unsupported features
+   *
+   * @param pattern Regex pattern
+   * @param flags Regex flags for interpreting special characters in the pattern
+   * @param capture Controls how capture groups in the pattern are used
+   * @return Instance of this object
+   */
+  static std::unique_ptr<regex_program> create(std::string_view pattern,
+                                               regex_flags flags      = regex_flags::DEFAULT,
+                                               capture_groups capture = capture_groups::EXTRACT);
+
+  /**
+   * @brief Move constructor
+   *
+   * @param other Object to move from
+   */
+  regex_program(regex_program&& other);
+
+  /**
+   * @brief Move operator assignment
+   *
+   * @param other Object to move from
+   * @return this object
+   */
+  regex_program& operator=(regex_program&& other);
+
+  /**
+   * @brief Return the pattern used to create this instance
+   *
+   * @return regex pattern as a string
+   */
+  std::string pattern() const;
+
+  /**
+   * @brief Return the regex_flags used to create this instance
+   *
+   * @return regex flags setting
+   */
+  regex_flags flags() const;
+
+  /**
+   * @brief Return the capture_groups used to create this instance
+   *
+   * @return capture groups setting
+   */
+  capture_groups capture() const;
+
+  /**
+   * @brief Return the number of instructions in this instance
+   *
+   * @return Number of instructions
+   */
+  int32_t instructions_count() const;
+
+  /**
+   * @brief Return the number of capture groups in this instance
+   *
+   * @return Number of groups
+   */
+  int32_t groups_count() const;
+
+  /**
+   * @brief Return implementation object
+   *
+   * @return impl object instance
+   */
+  regex_program_impl* get_impl() const;
```

Comment:
A few questions, possibly just due to my lack of experience with best practices with the pimpl idiom:
1. Should this method really be public? It looks like it's only ever used to do `create_prog_device`, maybe that should be a public method of `regex_program` instead? Otherwise it sort of breaks the pimpl encapsulation (although you still get the compilation benefits).
2. I'm not sure what `const` semantics this member's qualifier is really meant to imply. While it is true that it is technically const, it's also returning a non-const pointer to the program impl that could be modified. Of course, the implementation is also hidden behind an opaque pointer, so the calling code would have to know what the impl looks like, but it still seems unnecessarily insecure.
3. Could the returned value be a `regex_program_impl const*` instead? That would also seem safer (I haven't looked at the methods of `regex_program_impl` quite yet though so maybe is has to be a non-const object to allow calling non-const methods).

Reply:
I struggled with this a bit. I'm trying to hide the implementation. I need a way to convert a `regex_program` (public) into an internal `detail::reprog_device` object. Perhaps I need to turn it around somehow.
The `const` qualifier implies that the member function will not change any member for the instance it is called on.
The automatically implies the return pointer is const and you get a compile error if you try to declare it as such.

Reply:
I've reworked this to remove the `get_impl()` and used a friend class as an accessor instead. I think this looks alot cleaner and provides a better solution.

Reply:
The current solution looks good.